```
std/datetime/systime.d(87): Error: module `format` is in file 'std/format.d' which cannot be read
import path[0] = ../druntime/import
gmake: *** [posix.mak:554: checkwhitespace] Error 1
```
What?

--- a/src/dmd/astbase.d
+++ b/src/dmd/astbase.d
@@ -1562,7 +1562,7 @@ struct ASTBase
         {
             super(ident);
             this.arg = filename;
-            srcfile = FileName(FileName.defaultExt(filename.toDString, global.mars_ext));
+            srcfile = FileName(FileName.defaultExt(filename.toDString, mars_ext));
         }
 
         override void accept(Visitor v)

--- a/src/dmd/dmodule.d
+++ b/src/dmd/dmodule.d
@@ -55,6 +55,10 @@ version(Windows) {
 }
 
 /* ===========================  ===================== */
+
+enum package_d  = "package." ~ mars_ext;
+enum package_di = "package." ~ hdr_ext;
+
 /********************************************
  * Look for the source file if it's different from filename.
  * Look for .di, .d, directory, and along global.path.
@@ -69,11 +73,11 @@ private const(char)[] lookForSourceFile(const(char)[] filename)
 {
     /* Search along global.path for .di file, then .d file.
      */
-    const sdi = FileName.forceExt(filename, global.hdr_ext);
+    const sdi = FileName.forceExt(filename, hdr_ext);
     if (FileName.exists(sdi) == 1)
         return sdi;
     scope(exit) FileName.free(sdi.ptr);
-    const sd = FileName.forceExt(filename, global.mars_ext);
+    const sd = FileName.forceExt(filename, mars_ext);
     if (FileName.exists(sd) == 1)
         return sd;
     scope(exit) FileName.free(sd.ptr);
@@ -83,11 +87,11 @@ private const(char)[] lookForSourceFile(const(char)[] filename)
          * Therefore, the result should be: filename/package.d
          * iff filename/package.d is a file
          */
-        const ni = FileName.combine(filename, "package.di");
+        const ni = FileName.combine(filename, package_di);
         if (FileName.exists(ni) == 1)
             return ni;
         FileName.free(ni.ptr);
-        const n = FileName.combine(filename, "package.d");
+        const n = FileName.combine(filename, package_d);
         if (FileName.exists(n) == 1)
             return n;
         FileName.free(n.ptr);
@@ -114,11 +118,11 @@ private const(char)[] lookForSourceFile(const(char)[] filename)
         FileName.free(b.ptr);
         if (FileName.exists(n) == 2)
         {
-            const n2i = FileName.combine(n, "package.di");
+            const n2i = FileName.combine(n, package_di);
             if (FileName.exists(n2i) == 1)
                 return n2i;
             FileName.free(n2i.ptr);
-            const n2 = FileName.combine(n, "package.d");
+            const n2 = FileName.combine(n, package_d);
             if (FileName.exists(n2) == 1) {
                 return n2;
             }
@@ -531,7 +535,7 @@ extern (C++) final class Module : Package
         const(char)[] srcfilename;
         //printf("Module::Module(filename = '%s', ident = '%s')\n", filename, ident.toChars());
         this.arg = filename;
-        srcfilename = FileName.defaultExt(filename, global.mars_ext);
+        srcfilename = FileName.defaultExt(filename, mars_ext);
         if (target.run_noext && global.params.run &&
             !FileName.ext(filename) &&
             FileName.exists(srcfilename) == 0 &&
@@ -540,14 +544,14 @@ extern (C++) final class Module : Package
             FileName.free(srcfilename.ptr);
             srcfilename = FileName.removeExt(filename); // just does a mem.strdup(filename)
         }
-        else if (!FileName.equalsExt(srcfilename, global.mars_ext) &&
-                 !FileName.equalsExt(srcfilename, global.hdr_ext) &&
-                 !FileName.equalsExt(srcfilename, "dd"))
+        else if (!FileName.equalsExt(srcfilename, mars_ext) &&
+                 !FileName.equalsExt(srcfilename, hdr_ext) &&
+                 !FileName.equalsExt(srcfilename, dd_ext))
         {
 
             error("source file name '%.*s' must have .%.*s extension",
                   cast(int)srcfilename.length, srcfilename.ptr,
-                  cast(int)global.mars_ext.length, global.mars_ext.ptr);
+                  cast(int)mars_ext.length, mars_ext.ptr);
             fatal();
         }
 
@@ -556,7 +560,7 @@ extern (C++) final class Module : Package
         if (doDocComment)
             setDocfile();
         if (doHdrGen)
-            hdrfile = setOutfilename(global.params.hdrname, global.params.hdrdir, arg, global.hdr_ext);
+            hdrfile = setOutfilename(global.params.hdrname, global.params.hdrdir, arg, hdr_ext);
         escapetable = new Escape();
     }
 
@@ -678,7 +682,7 @@ extern (C++) final class Module : Package
 
     extern (D) void setDocfile()
     {
-        docfile = setOutfilename(global.params.docname, global.params.docdir, arg, global.doc_ext);
+        docfile = setOutfilename(global.params.docname, global.params.docdir, arg, doc_ext);
     }
 
     /**
@@ -710,7 +714,7 @@ extern (C++) final class Module : Package
         else
         {
             // if module is not named 'package' but we're trying to read 'package.d', we're looking for a package module
-            bool isPackageMod = (strcmp(toChars(), "package") != 0) && (strcmp(srcfile.name(), "package.d") == 0 || (strcmp(srcfile.name(), "package.di") == 0));
+            bool isPackageMod = (strcmp(toChars(), "package") != 0) && (strcmp(srcfile.name(), package_d) == 0 || (strcmp(srcfile.name(), package_di) == 0));
             if (isPackageMod)
                 .error(loc, "importing package '%s' requires a 'package.d' file which cannot be found in '%s'", toChars(), srcfile.toChars());
             else
@@ -894,8 +898,8 @@ extern (C++) final class Module : Package
 
         const(char)* srcname = srcfile.toChars();
         //printf("Module::parse(srcname = '%s')\n", srcname);
-        isPackageFile = (strcmp(srcfile.name(), "package.d") == 0 ||
-                         strcmp(srcfile.name(), "package.di") == 0);
+        isPackageFile = (strcmp(srcfile.name(), package_d) == 0 ||
+                         strcmp(srcfile.name(), package_di) == 0);
         const(char)[] buf = cast(const(char)[]) srcBuffer.data;
 
         bool needsReencoding = true;
@@ -1016,7 +1020,7 @@ extern (C++) final class Module : Package
          * but do not have to if they have the .dd extension.
          * https://issues.dlang.org/show_bug.cgi?id=15465
          */
-        if (FileName.equalsExt(arg, "dd"))
+        if (FileName.equalsExt(arg, dd_ext))
         {
             comment = buf.ptr; // the optional Ddoc, if present, is handled above.
             isDocFile = true;
@@ -1026,7 +1030,7 @@ extern (C++) final class Module : Package
         }
         /* If it has the extension ".di", it is a "header" file.
          */
-        if (FileName.equalsExt(arg, "di"))
+        if (FileName.equalsExt(arg, hdr_ext))
         {
             isHdrFile = true;
         }
@@ -1068,8 +1072,8 @@ extern (C++) final class Module : Package
             }
             assert(dst);
             Module m = ppack ? ppack.isModule() : null;
-            if (m && (strcmp(m.srcfile.name(), "package.d") != 0 &&
-                      strcmp(m.srcfile.name(), "package.di") != 0))
+            if (m && (strcmp(m.srcfile.name(), package_d) != 0 &&
+                      strcmp(m.srcfile.name(), package_di) != 0))
             {
                 .error(md.loc, "package name '%s' conflicts with usage as a module name in file %s", ppack.toPrettyChars(), m.srcfile.toChars());
             }

--- a/src/dmd/doc.d
+++ b/src/dmd/doc.d
@@ -361,7 +361,7 @@ private TemplateDeclaration getEponymousParent(Dsymbol s)
     return (td && getEponymousMember(td)) ? td : null;
 }
 
-private immutable ddoc_default = import("default_ddoc_theme.ddoc");
+private immutable ddoc_default = import("default_ddoc_theme." ~ ddoc_ext);
 private immutable ddoc_decl_s = "$(DDOC_DECL ";
 private immutable ddoc_decl_e = ")\n";
 private immutable ddoc_decl_dd_s = "$(DDOC_DECL_DD ";

--- a/src/dmd/frontend.h
+++ b/src/dmd/frontend.h
@@ -7279,13 +7279,6 @@ typedef uint32_t structalign_t;
 struct Global
 {
     _d_dynamicArray< const char > inifilename;
-    _d_dynamicArray< const char > mars_ext;
-    _d_dynamicArray< const char > doc_ext;
-    _d_dynamicArray< const char > ddoc_ext;
-    _d_dynamicArray< const char > hdr_ext;
-    _d_dynamicArray< const char > cxxhdr_ext;
-    _d_dynamicArray< const char > json_ext;
-    _d_dynamicArray< const char > map_ext;
     _d_dynamicArray< const char > copyright;
     _d_dynamicArray< const char > written;
     Array<const char* >* path;
@@ -7311,13 +7304,6 @@ struct Global
     ~Global();
     Global() :
         inifilename(),
-        mars_ext(1, "d"),
-        doc_ext(4, "html"),
-        ddoc_ext(4, "ddoc"),
-        hdr_ext(2, "di"),
-        cxxhdr_ext(1, "h"),
-        json_ext(4, "json"),
-        map_ext(3, "map"),
         copyright(73, "Copyright (C) 1999-2021 by The D Language Foundation, All Rights Reserved"),
         written(24, "written by Walter Bright"),
         path(),
@@ -7334,15 +7320,8 @@ struct Global
         debugids()
     {
     }
-    Global(_d_dynamicArray< const char > inifilename, _d_dynamicArray< const char > mars_ext = { 1, "d" }, _d_dynamicArray< const char > doc_ext = { 4, "html" }, _d_dynamicArray< const char > ddoc_ext = { 4, "ddoc" }, _d_dynamicArray< const char > hdr_ext = { 2, "di" }, _d_dynamicArray< const char > cxxhdr_ext = { 1, "h" }, _d_dynamicArray< const char > json_ext = { 4, "json" }, _d_dynamicArray< const char > map_ext = { 3, "map" }, _d_dynamicArray< const char > copyright = { 73, "Copyright (C) 1999-2021 by The D Language Foundation, All Rights Reserved" }, _d_dynamicArray< const char > written = { 24, "written by Walter Bright" }, Array<const char* >* path = nullptr, Array<const char* >* filePath = nullptr, _d_dynamicArray< const char > vendor = {}, Param params = Param(true, true, false, false, false, false, false, false, false, false, false, false, false, false, false, false, true, 0u, false, false, (DiagnosticReporting)1u, false, false, false, (FeatureState)-1, false, false, false, (DiagnosticReporting)2u, (PIC)0u, false, false, 0u, false, false, false, true, true, true, false, false, false, false, false, false, false, false, false, false, false, (FeatureState)-1, false, false, (CppStdRevision)201103u, true, false, false, false, false, false, false, false, false, false, false, false, false, false, false, (CHECKENABLE)0u, (CHECKENABLE)0u, (CHECKENABLE)0u, (CHECKENABLE)0u, (CHECKENABLE)0u, (CHECKENABLE)0u, (CHECKENABLE)0u, (CHECKACTION)0u, 20u, {}, Array<const char* >(0LLU, {}, arrayliteral), nullptr, nullptr, {}, {}, {}, false, {}, {}, Array<const char* >(0LLU, {}, arrayliteral), false, {}, {}, true, (CxxHeaderMode)0u, {}, {}, false, {}, (JsonFieldFlags)0u, nullptr, nullptr, 0, 0u, nullptr, 0u, nullptr, {}, {}, {}, {}, nullptr, false, {}, Array<const char* >(0LLU, {}, arrayliteral), (MessageStyle)0u, false, Array<const char* >(0LLU, {}, arrayliteral), Array<const char* >(0LLU, {}, arrayliteral), Array<const char* >(0LLU, {}, arrayliteral), Array<bool >(0LLU, {}, arrayliteral), Array<const char* >(0LLU, {}, arrayliteral), Array<const char* >(0LLU, {}, arrayliteral), {}, {}, {}, {}), uint32_t errors = 0u, uint32_t warnings = 0u, uint32_t gag = 0u, uint32_t gaggedErrors = 0u, uint32_t gaggedWarnings = 0u, void* console = nullptr, Array<Identifier* >* versionids = nullptr, Array<Identifier* >* debugids = nullptr) :
+    Global(_d_dynamicArray< const char > inifilename, _d_dynamicArray< const char > copyright = { 73, "Copyright (C) 1999-2021 by The D Language Foundation, All Rights Reserved" }, _d_dynamicArray< const char > written = { 24, "written by Walter Bright" }, Array<const char* >* path = nullptr, Array<const char* >* filePath = nullptr, _d_dynamicArray< const char > vendor = {}, Param params = Param(true, true, false, false, false, false, false, false, false, false, false, false, false, false, false, false, true, 0u, false, false, (DiagnosticReporting)1u, false, false, false, (FeatureState)-1, false, false, false, (DiagnosticReporting)2u, (PIC)0u, false, false, 0u, false, false, false, true, true, true, false, false, false, false, false, false, false, false, false, false, false, (FeatureState)-1, false, false, (CppStdRevision)201103u, true, false, false, false, false, false, false, false, false, false, false, false, false, false, false, (CHECKENABLE)0u, (CHECKENABLE)0u, (CHECKENABLE)0u, (CHECKENABLE)0u, (CHECKENABLE)0u, (CHECKENABLE)0u, (CHECKENABLE)0u, (CHECKACTION)0u, 20u, {}, Array<const char* >(0LLU, {}, arrayliteral), nullptr, nullptr, {}, {}, {}, false, {}, {}, Array<const char* >(0LLU, {}, arrayliteral), false, {}, {}, true, (CxxHeaderMode)0u, {}, {}, false, {}, (JsonFieldFlags)0u, nullptr, nullptr, 0, 0u, nullptr, 0u, nullptr, {}, {}, {}, {}, nullptr, false, {}, Array<const char* >(0LLU, {}, arrayliteral), (MessageStyle)0u, false, Array<const char* >(0LLU, {}, arrayliteral), Array<const char* >(0LLU, {}, arrayliteral), Array<const char* >(0LLU, {}, arrayliteral), Array<bool >(0LLU, {}, arrayliteral), Array<const char* >(0LLU, {}, arrayliteral), Array<const char* >(0LLU, {}, arrayliteral), {}, {}, {}, {}), uint32_t errors = 0u, uint32_t warnings = 0u, uint32_t gag = 0u, uint32_t gaggedErrors = 0u, uint32_t gaggedWarnings = 0u, void* console = nullptr, Array<Identifier* >* versionids = nullptr, Array<Identifier* >* debugids = nullptr) :
         inifilename(inifilename),
-        mars_ext(mars_ext),
-        doc_ext(doc_ext),
-        ddoc_ext(ddoc_ext),
-        hdr_ext(hdr_ext),
-        cxxhdr_ext(cxxhdr_ext),
-        json_ext(json_ext),
-        map_ext(map_ext),
         copyright(copyright),
         written(written),
         path(path),

--- a/src/dmd/globals.d
+++ b/src/dmd/globals.d
@@ -261,16 +261,17 @@ alias structalign_t = uint;
 // other values are all powers of 2
 enum STRUCTALIGN_DEFAULT = (cast(structalign_t)~0);
 
+enum mars_ext = "d";
+enum doc_ext  = "html";     // for Ddoc generated files
+enum ddoc_ext = "ddoc";     // for Ddoc macro include files
+enum dd_ext   = "dd";       // for Ddoc source files
+enum hdr_ext  = "di";       // for D 'header' import files
+enum json_ext = "json";     // for JSON files
+enum map_ext  = "map";      // for .map files
+
 extern (C++) struct Global
 {
     const(char)[] inifilename;
-    string mars_ext = "d";
-    string doc_ext = "html";      // for Ddoc generated files
-    string ddoc_ext = "ddoc";     // for Ddoc macro include files
-    string hdr_ext = "di";        // for D 'header' import files
-    string cxxhdr_ext = "h";      // for C/C++ 'header' files
-    string json_ext = "json";     // for JSON files
-    string map_ext = "map";       // for .map files
 
     string copyright = "Copyright (C) 1999-2021 by The D Language Foundation, All Rights Reserved";
     string written = "written by Walter Bright";

--- a/src/dmd/globals.h
+++ b/src/dmd/globals.h
@@ -243,17 +243,17 @@ typedef unsigned structalign_t;
 // other values are all powers of 2
 #define STRUCTALIGN_DEFAULT ((structalign_t) ~0)
 
+const DString mars_ext = "d";
+const DString doc_ext  = "html";     // for Ddoc generated files
+const DString ddoc_ext = "ddoc";     // for Ddoc macro include files
+const DString dd_ext   = "dd";       // for Ddoc source files
+const DString hdr_ext  = "di";       // for D 'header' import files
+const DString json_ext = "json";     // for JSON files
+const DString map_ext  = "map";      // for .map files
+
 struct Global
 {
     DString inifilename;
-    const DString mars_ext;
-    const DString doc_ext;      // for Ddoc generated files
-    const DString ddoc_ext;     // for Ddoc macro include files
-    const DString hdr_ext;      // for D 'header' import files
-    const DString cxxhdr_ext;   // for C/C++ 'header' files
-    const DString json_ext;     // for JSON files
-    const DString map_ext;      // for .map files
-
 
     const DString copyright;
     const DString written;

--- a/src/dmd/link.d
+++ b/src/dmd/link.d
@@ -216,7 +216,7 @@ public int runLINK()
 
     const(char)[] getMapFilename()
     {
-        const(char)[] fn = FileName.forceExt(global.params.exefile, "map");
+        const(char)[] fn = FileName.forceExt(global.params.exefile, map_ext);
         const(char)[] path = FileName.path(global.params.exefile);
         return path.length ? fn : FileName.combine(global.params.objdir, fn);
     }
@@ -593,7 +593,7 @@ public int runLINK()
             }
             if (!global.params.mapfile.length)
             {
-                const(char)[] fn = FileName.forceExt(global.params.exefile, "map");
+                const(char)[] fn = FileName.forceExt(global.params.exefile, map_ext);
                 const(char)[] path = FileName.path(global.params.exefile);
                 global.params.mapfile = path.length ? fn : FileName.combine(global.params.objdir, fn);
             }

--- a/src/dmd/mars.d
+++ b/src/dmd/mars.d
@@ -919,7 +919,7 @@ extern (C++) void generateJson(Modules* modules)
         const(char)[] jsonfilename;
         if (name)
         {
-            jsonfilename = FileName.defaultExt(name, global.json_ext);
+            jsonfilename = FileName.defaultExt(name, json_ext);
         }
         else
         {
@@ -933,7 +933,7 @@ extern (C++) void generateJson(Modules* modules)
             n = FileName.name(n);
             //if (!FileName::absolute(name))
             //    name = FileName::combine(dir, name);
-            jsonfilename = FileName.forceExt(n, global.json_ext);
+            jsonfilename = FileName.forceExt(n, json_ext);
         }
         writeFile(Loc.initial, jsonfilename, buf[]);
     }
@@ -2420,7 +2420,7 @@ bool parseCommandLine(const ref Strings arguments, const size_t argc, ref Param 
             {
                 const(char)[] runarg = arguments[i + 1].toDString();
                 const(char)[] ext = FileName.ext(runarg);
-                if (ext && FileName.equals(ext, "d") == 0 && FileName.equals(ext, "di") == 0)
+                if (ext && FileName.equals(ext, mars_ext) == 0 && FileName.equals(ext, hdr_ext) == 0)
                 {
                     error("-run must be followed by a source file, not '%s'", arguments[i + 1]);
                     break;
@@ -2718,18 +2718,18 @@ Module createModule(const(char)* file, ref Strings libmodules)
             return null;
         }
     }
-    if (ext == global.ddoc_ext)
+    if (ext == ddoc_ext)
     {
         global.params.ddocfiles.push(file);
         return null;
     }
-    if (FileName.equals(ext, global.json_ext))
+    if (FileName.equals(ext, json_ext))
     {
         global.params.doJsonGeneration = true;
         global.params.jsonfilename = file.toDString;
         return null;
     }
-    if (FileName.equals(ext, global.map_ext))
+    if (FileName.equals(ext, map_ext))
     {
         global.params.mapfile = file.toDString;
         return null;
@@ -2754,7 +2754,7 @@ Module createModule(const(char)* file, ref Strings libmodules)
     /* Examine extension to see if it is a valid
         * D source file extension
         */
-    if (FileName.equals(ext, global.mars_ext) || FileName.equals(ext, global.hdr_ext) || FileName.equals(ext, "dd"))
+    if (FileName.equals(ext, mars_ext) || FileName.equals(ext, hdr_ext) || FileName.equals(ext, dd_ext))
     {
         name = FileName.removeExt(p);
         if (!name.length || name == ".." || name == ".")

